active/0000.prebuild-addon.md

```diff
@@ -0,0 +1,111 @@
+
+
+## Trees that are safe to prebuild
+
+By default, the templates, addon, addon-test-support, vendor trees will be prebuilt. A tree can be safely prebuilt if the `treeFor` hooks for the said trees are not being extended by the addon. If the hook has been extended, it is the responsibility of the addon owner to decide if the tree is safe to prebuilt else opt out of prebuilding for any tree by specifying it in the index.js of the addon.
+
+As a next step, `ember-cli-prebuild-addon` should be able to figure out if the addon has extended `treeFor` hooks and opt out the respective trees from being prebuilt.
+
+## Location to store the prebuilt addons
+
+The addon can specify the location the prebuilt directory can be stored in its index.js. By default, the prebuilt directory will be stored inside the addon whether the addon is prebuilt by the app or the addon itself.
+
+The prebuilt directory will contain the prebuilt addon one for each target. Md5 checksum of each target file is used to generate the directory name.
+
+The checksum will be calculated by sorting the array of browsers to ensure stability.
+
+It will in the format `pre-built/<md5ChecksumOfTargetgroup-addonVersion>/addon`. 
+Ember-cli will check if the addon is prebuilt for the same target group by calculating the md5 checksum of app's targets.
+
+## Configuring Target Groups
+
+The addons should be prebuilt for multiple target group. A target group will contain a list of browsers. Prebuilding for multiple target group is necessary because if the application developer decides to use only chrome for building the app then shipping the build code for other browsers to the client is unnecessary and will also impact the build time significantly. Hence having the flexibility to be able to prebuild for different target group is highly important.
+
+Currently `ember prebuild` will look for target group in 
+
+-`<configPath>/prebuild` which can contain a list of target files. The package.json of the addon or app should be able to specify the list of target files that should be used to prebuild. 
+-If no target files are present in `config/prebuild`, then it will look for `<configPath>/target.js` 
+-If the above ones are not successful, the default ember-cli target will be used.
+
+## Changes in Addon BluePrint
+
+For the addons to be prebuilt by default,
+
+-The addons should have `ember-cli-prebuild-addon` in its package.json
+-`ember prebuild` command should be included in  npm `prepack` hook so that the addon will prebuilt before publishing to npm registry.
+
+# Drawbacks
+
+If the application is using the incorrect prebuilt addon, the application build will break
+
+-The latest version of target browser changes continuously hence the prebuilt directory might not necessarily be built for the latest version of the browser. But since current last version will always be a subset of future last versions the prebuilt directory will always be valid. 
```

Comment:
> But since current last version will always be a subset of future last versions the prebuilt directory will always be valid.

This assertion is not correct. Features have been removed from browsers (eg. `SharedArrayBuffer` due to Spectre).

You also assume that Babel / preset-env is infallible, and that the generated output is always correct. This is _NOT_ the case. If a transpilation issue arises, it will no longer be fixed with an upstream update, but instead requires the ember addon to publish an update, though no addon code has a changed.